enhancements/installer/aws-custom-dns.md

```diff
@@ -0,0 +1,488 @@
+DNS with entries for API, Internal API and Ingress Services. 
+
+2. The Installer and the Ingress Operator continue to configure the cloud LBs
+on the public subnets of Public Hosted Zones for `api` and `*.apps`. They will
+continue to configure cloud LBs on the private subnets of Private Hosted Zones
+for `api-int` and `*.apps` for in-cluster access to the Ingress service.
+
+3. The Installer will also update the new fields `APIServerDNSConfig` and
+`InternalAPIServerDNSConfig` within the `AWSPlatformStatus`, with the LB
+IP addresses and the DNS Record types for API, API-Int servers respectively.
+
+4. The Installer instructs the machine-config-operator(MCO) to start a static
+CoreDNS pod on the bootstrap and master nodes to provide DNS resolution for
+the API and API-Int servers. MCO uses the DNS config in the `AWSPlatformStatus`
+to generate the CoreFile to configure the CoreDNS instance in the pod.
+Note: This implementation differs from the on-prem case by not staring the
+keepalived or haproxy containers for load balancing.
+
+5. The Installer omits the DNS Zones within `dnses.config.openshift.io/cluster`
+to influence the behavior of the Ingress Operator by making it skip the
+configuration of cloud DNS while continuing with LB configuration.
+
+6. The Ingress Controllers managed by the cluster-ingress-operator continue
+to configure the application LBs to enable in-cluster and external accesss to
+cluster services. The Ingress Controllers will continue to update the DNSRecord
+ConfigResource(CR) with the configuration required to configure DNS for these
+services.
+
+7. The MCO will watch the DNSRecord CR for the DNS configuration mentioned in
+step 6. MCO will update the newly added `IngressDNSConfig` field within the
+`AWSPlatformStatus` with the DNS configuartion for *.apps records by reading
+the DNSRecord CR.
+
+8. The runtimecfg utility will monitor the `Infrastructure` CR for changes to
+`IngressDNSConfig` of `AWSPlatformStatus` to update the CoreDNS pod's
+CoreFile to now also include DNS configuration for both internal and
+external access to the cluster services (the `*.apps` records).
+
+9. Once the cluster is functional, the user can look at the contents of the new
+fields `APIServerDNSConfig` and `IngressDNSConfig` in the `AWSPlatformStatus`
+field of the `Infrastructure` CR to configure their external DNS with entries
+for `api` and `*.apps` service.
+
+### API Extensions
+
+1. The AWSPlatformStatus within the PlatformStatus field of the Infrastructure
+ConfigResource (CR) is updated to contain DNS config required to configure
+the in-cluster CoreDNS solution. This same CR is available to the user post
+a successful cluster install, to configure their own DNS solution.
+
+```go
+type AWSPlatformStatus struct {
+        <snip>
+        // APIServerDNSConfig contains information to configure DNS for API Server.
+        // This field will be set only when the userConfiguredDNS feature is enabled.
```

Comment:
No. Current understanding is that having non-empty DNSConfig entries within AWSPlatformStatus will be sufficient. The user sets the `userConfiguredDNS` field in the install-config to Enabled to turn on this feature.

Reply:
That's an implicit behaviour, I wonder if we want something more explicit/declarative?

If not, this comment should be updated because `userConfiguredDNS` isn't a thing in the cluster, I'm thinking it would be more like `The presence of this field allows configuration of external DNS` or something like that, more about the long term feature from a user perspective than a specific installer feature, WDYT?

Reply:
Regarding the comment, I was trying to clarify that ClusterDNSConfig will contain non-empty values when a specific action is triggered by the user (setting the userConfiguredDNS config specifically to Enabled) and not just as a result of the general running of the cluster. Is the API review https://github.com/openshift/api/pull/1397 a better place to discuss this?

Regarding, adding an explicit config item to indicate custom DNS is in use, I did consider that while designing this. I was under the impression that we would want to keep the number of net new config items to the minimum so if there were other ways to get this information, we should utilize those. No strong preference. Is this a good item to has out in the implementation review?

Reply:
Lets continue the discussion on the API PR then